DigitalLearningSolutions.Web/ServiceFilter/ValidateAllowedDlsSubApplication.cs

```diff
@@ -33,42 +37,83 @@ public async void OnActionExecuting(ActionExecutingContext context)
             var user = context.HttpContext.User;
             if (!user.Identity.IsAuthenticated)
             {
-                RedirectToLogin(context);
+                SetLoginRedirectResult(context);
                 return;
             }
 
-            var dlsSubApplicationParameterName = context.ActionDescriptor.Parameters?
+            if (!TryGetDlsSubApplication(context, out var application, out var applicationParameterName))
+            {
+                return;
+            }
+
+            if (await ApplicationIsInaccessibleByPage(context, application))
+            {
+                return;
+            }
+
+            ValidateUserHasPermissionForApplication(context, user, application, applicationParameterName!);
+        }
+
+        private bool TryGetDlsSubApplication(
+            ActionExecutingContext context,
+            out DlsSubApplication? application,
+            out string? dlsSubApplicationParameterName
+        )
+        {
+            dlsSubApplicationParameterName = context.ActionDescriptor.Parameters?
                 .FirstOrDefault(x => x.ParameterType == typeof(DlsSubApplication))?.Name;
 
-            if (!string.IsNullOrEmpty(dlsSubApplicationParameterName) &&
-                HasModelBindingError(context, dlsSubApplicationParameterName!))
+            var actionDoesNotRequireApplication = dlsSubApplicationParameterName == null;
+            if (actionDoesNotRequireApplication)
+            {
+                application = null;
+                return false;
+            }
+
+            if (HasModelBindingError(context, dlsSubApplicationParameterName!))
             {
                 SetNotFoundResult(context);
-                return;
+                application = null;
+                return false;
             }
 
-            var application = dlsSubApplicationParameterName == null
-                ? null
-                : GetDlsSubApplicationFromContext(context, dlsSubApplicationParameterName);
+            application = GetDlsSubApplicationFromContext(context, dlsSubApplicationParameterName!);

```

Comment:
Does this definitely give us a non-null value for application? I think it would be nice to rule out the null-application case when we call this method.
E.g. if the application param on the action is nullable or optional.

Reply:
There's a check that the actual arguments actually contains the key the actionDescription claims they do, but I don't know if there's a way to stop the value attached to that key being null. Maybe the model binder would have pitched a fit at that point, however.

Reply:
Okay, well I think it's fine to return a nullable application if we're not sure, but I think we should return a NotFound response if it's null after we call `TryGetDlsSubApplication()`, because I don't think this attribute should ever be used by an endpoint that isn't expecting some non-null dlsSubApplication. There are only a few endpoints that aren't part of an sub-application (if any), and it's very unlikely any of them will need to appear with other sub-application nav bars .